plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs

```diff
@@ -0,0 +1,317 @@
+  -- ^ A custom user-controlled generator for terms - useful for situations where
+  -- we want to e.g. generate custom strings for coverage or test some specific
+  -- pattern that generates a special case for the compiler.
+  , geCustomFreq         :: Int
+  -- ^ How often do we use the custom generator -
+  -- values in the range of 10-30 are usually reasonable.
+  , geDebug              :: Bool
+  -- ^ Are we currently running in debug-mode (to debug our generators)
+  }
+
+{- Note [Recursion Control and geSize]
+One would be forgiven for thinking that you don't need `geSize` in `GenTm` because size is built-in
+to 'Gen'. However, if you use 'Gen's built-in size to control the size of both the terms you
+generate *and* the size of the constants in the terms you will end up with skewed terms. Constants
+near the top of the term will be big and constants near the bottom of the term will be small. For
+this reason we follow QuickCheck best practise and keep track of the "recursion control size"
+separately from 'Gen's size in the 'geSize' field of the 'GenEnv' environment. I.e. we let the
+QuickCheck's size parameter to be responsible for the size of constants at the leaves of the AST
+and use 'geSize' to control the size of the AST itself.
+-}
+
+-- | Run a generator in debug-mode.
+withDebug :: GenTm a -> GenTm a
+withDebug gen = local (\env -> env { geDebug = True }) gen
+
+-- | Run a `GenTm  generator in a top-level empty context where we are allowed to generate
+-- datatypes.
+runGenTm :: GenTm a -> Gen a
+runGenTm = runGenTmCustom 0 (error "No custom generator.")
+
+-- | Run a `GenTm` generator with a plug-in custom generator for terms that is included with
+-- the other generators.
+runGenTmCustom :: Int
+               -> (Maybe (Type TyName DefaultUni ())
+                  -> GenTm (Type TyName DefaultUni (), Term TyName Name DefaultUni DefaultFun ()))
+               -> GenTm a
+               -> Gen a
+runGenTmCustom f cg g = do
+  sized $ \ n -> do
+    let env = GenEnv
+          { geSize               = n
+          , geDatas              = Map.empty
+          , geTypes              = Map.empty
+          , geTerms              = Map.empty
+          , geUnboundUsedTyNames = Set.empty
+          , geEscaping           = YesEscape
+          , geCustomGen          = cg
+          , geCustomFreq         = f
+          , geDebug              = False
+          }
+    flip runReader env <$> runGenT g
+
+suchThatMap :: Monad m => GenT m a -> (a -> Maybe b) -> GenT m b
+gen `suchThatMap` f = fmap fromJust $ fmap f gen `suchThat` isJust
+
+deliver :: Monad m => GenT m (Maybe a) -> GenT m a
+deliver gen = gen `suchThatMap` id
+
+-- * Utility functions
+
+-- | Don't allow types to escape from a generator.
+withNoEscape :: GenTm a -> GenTm a
+withNoEscape = local $ \env -> env { geEscaping = NoEscape }
+
+-- * Dealing with size
+
+-- | Map a function over the generator size
```

Comment:
> I read this and I was immediately confused. I think it's maybe a mistake to use the same name "size" for both the quickcheck size and geSize. Could we call the recursion size "recursion depth" or something? Otherwise I'm going to get confused every time I see one of these size functions.

Depth is a very different thing. E.g. one of the major differences between NEAT and SmallCheck is that the former uses size and the latter uses depth. I get your point though, made it `*AstSize` instead of `*Size` everywhere.